### DISCUSSION TITLE
Sync Sharepoint site Documents

### DISCUSSION THREAD

**Discussion by hugobrh** [abraunegg/onedrive#2258] [Q&A] [answered]
Hello and thank you for the work!
I was wondering if it is possible to synchronize a sharepoint site's shared documents. Apparently, it is considered as the site library ? I don't really understand all about Onedrive-Sharepoint. Let's say I have a site called Test. I tried to sync via:
`onedrive --get-O365-drive-id 'Test'` or `onedrive --get-O365-drive-id 'sites/Test'`
But it returns :
```
Using dry-run database copy for OneDrive API query
Configuring Global Azure AD Endpoints
Initializing the Synchronization Engine ...
Office 365 Library Name Query: sites/Test_HB
ERROR: Query of OneDrive for Office 365 Library Name failed
ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 404 (Not Found)
  Error Reason:     Item not found
```
By the way I'm on Ubuntu 22.04.

**Comment by abraunegg** [comment, marked as answer]
@hugobrehier 
> By the way I'm on Ubuntu 22.04.

Firstly what application version are you using? Please ensure you are running v2.4.22 or greater.

If a site name cannot be found, the following application output is performed:
```
./onedrive --confdir '~/.config/onedrive-business' --get-O365-drive-id 'sites/Test'
Configuration file successfully loaded
Using dry-run database copy for OneDrive API query
Configuring Global Azure AD Endpoints
Initializing the Synchronization Engine ...
Office 365 Library Name Query: sites/Test

ERROR: The requested SharePoint site could not be found. Please check it's name and your permissions to access the site.

The following SharePoint site names were returned:
 * testteamsite
 * Communication site
 * Data Production
 * testteamsite2
 * test site
 * All Company
 * PointPublishing Hub Site
 * Community
 * Team Site
```

However your error message is being generated from the initial Site Search - so this is either a 'configuration' issue with how you have configured the client and/or a permissions issue.

Things for you to do:
* Firstly because you chose Ubuntu as your platform, please ensure you are using the OpenSuSE Build Service repository to install the client - https://github.com/abraunegg/onedrive/blob/master/docs/ubuntu-package-install.md
* Ensure you have the correct application permissions to query SharePoint. Potentially use `--reauth` and ensure that you have appropriate permissions requested and used. You need to use your 'corporate|business' OneDrive Account - a 'personal' account will not work.
  * You may even need to get your corporate|organisation to update your permissions so that you can access this OneDrive API: https://learn.microsoft.com/en-us/onedrive/developer/rest-api/api/site_search?view=odsp-graph-online
* You may need to configure a specific 'azure_tenant_id' as well - please speak to whomever supports your SharePoint environment as to if this is the case. It is rare that this needs to be configured as well.

To me this sounds like a permissions issue|account issue - as the API cannot be found, which typically means no access.

**Reply by hugobrh** [reply]
I did install via the instructions linked in your first bullet point and my version is onedrive v2.4.22-1+np1
If I do a dry run, I do have results : 
```
Using 'user' Config Dir: /home/xxx/.config/onedrive
Using 'system' Config Dir: /etc/onedrive
No user or system config file found, using application defaults
DRY-RUN Configured. Output below shows what 'would' have occurred.
Checking Application Version ...
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Using Curl defaults for all HTTP operations
Opening the item database ...
All operations will be performed in: /home/xxx/OneDrive
Application version: v2.4.22-1+np1
Account Type: business
Default Drive ID: xxx
Default Root ID: xxx
Remaining Free Space: xxx
Fetching details for OneDrive Root
OneDrive Root exists in the database
Initializing the Synchronization Engine ...
Syncing changes and items from OneDrive ...
Applying changes of Path ID: xxx
Updated Remaining Free Space: xxx
Number of items from OneDrive to process: 3
Downloading file xxx.pdf ... done.
Performing a database consistency and integrity check on locally stored data ... 
Uploading differences of ~/OneDrive
Processing .
The directory has not changed
Processing xxx
The directory has not changed
Processing xxx.pdf
The file has not changed
Uploading new items of ~/OneDrive
Applying changes of Path ID: xxx
Updated Remaining Free Space: xxx
Sync with OneDrive is complete
```

In fact, the synced folder is only a **non**-shared one.
I guess I don't have access.
I'll still be able to sync a personal OD however, thank you!

**Comment by rafalcode** [comment]
I'm having the same error output as hugobrehier, and my opinion is that the query is not successful here, operationally speaking.
I.e. it doesn't get to start to look for something, never mind detecting that it is not there.

I'm on Debian and onedrive syncing works fine, as well as business shared ... all good (thanks!). On the web interface, I can well see the sharepoint "libraries" (of course when logged in ...can create, edit and save), but the onedrive client ....no name that I can see in the web interface works.

Also I thought the whole idea was that this is for setting up the configuration, yet abraunegg uses a already set up configuration, so it's step ahead. If access was being denied, I'd really like to get a more explicit confirmation of that.

**Comment by abraunegg** [comment]
@rafalcode 
> Also I thought the whole idea was that this is for setting up the configuration, yet abraunegg uses a already set up configuration, so it's step ahead.

Answering this first ... I have a different configuration file / folder setup for when testing things - so that, if any data downloads it goes to that folder.

In your situation - you may be using your currently authenticated client - which will be your OneDrive Business Account ... you may, or may not have configured that to use a separate configuration directory ... the point is - all I am doing here is passing in a configuration directory, where that directory holds the authentication data for the client.

I can even run the *same* command (as below) on a totally 'clean' install:

```
[alex@onedrive-client-dev onedrive]$ ./onedrive --get-O365-drive-id '*'
Using dry-run database copy for OneDrive API query
Configuring Global Azure AD Endpoints
Authorize this app visiting:

https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=d50ca740-c83f-4d1b-b616-12c519384f0c&scope=Files.ReadWrite%20Files.ReadWrite.All%20Sites.ReadWrite.All%20offline_access&response_type=code&prompt=login&redirect_uri=https://login.microsoftonline.com/common/oauth2/nativeclient

Enter the response uri: https://login.microsoftonline.com/common/oauth2/nativeclient?code=<redacted>
Initializing the Synchronization Engine ...
Office 365 Library Name Query: *

ERROR: The requested SharePoint site could not be found. Please check it's name and your permissions to access the site.

The following SharePoint site names were returned:
 * testteamsite
 * Communication site
 * Data Production
 * testteamsite2
 * mynasau3
 * test site
 * All Company
 * PointPublishing Hub Site
 * Community
 * Team Site
[alex@onedrive-client-dev onedrive]$ 

```

The point here is - you *must* first use a client that is authenticated with your Business Account, and have the right permissions assigned to you by your organisation before you use `--get-O365-drive-id`

Now .. if I do the exact same test, but use a OneDrive Personal Account, this is the response I get:
```
[alex@onedrive-client-dev onedrive]$ ./onedrive --get-O365-drive-id '*'
Using dry-run database copy for OneDrive API query
Configuring Global Azure AD Endpoints
Authorize this app visiting:

https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=d50ca740-c83f-4d1b-b616-12c519384f0c&scope=Files.ReadWrite%20Files.ReadWrite.All%20Sites.ReadWrite.All%20offline_access&response_type=code&prompt=login&redirect_uri=https://login.microsoftonline.com/common/oauth2/nativeclient

Enter the response uri: https://login.microsoftonline.com/common/oauth2/nativeclient?code=<redacted>
Initializing the Synchronization Engine ...
Office 365 Library Name Query: *
ERROR: Query of OneDrive for Office 365 Library Name failed

ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 500 (Internal Server Error)
  Error Reason:     Unable to find target address
  Error Timestamp:  2022-12-13T21:22:41
  API Request ID:   8fbe24b2-7178-428e-a1a6-c523f72117b8
[alex@onedrive-client-dev onedrive]$ 
```


> I'm on Debian and onedrive syncing works fine

OK ... however you are on Debian. What application version are you using - what is the output of `onedrive --version` ? How did you install the client - did you follow: https://github.com/abraunegg/onedrive/blob/master/docs/ubuntu-package-install.md ?

> but the onedrive client ....no name that I can see in the web interface works.

This will be because your SharePoint site is not the name you 'think' it is. Is the site actually a SharePoint site?

Try this command:

```
onedrive --get-O365-drive-id '*'
```

This should return *any* SharePoint sites that you have access to.

If this does not, please provide the full error + a screen shot of the SharePoint Sites you are trying to access.

**Comment by abraunegg** [comment]
@hugobrh  , @rafalcode 
Whilst I am unable to reproduce this exact error that you both are getting (without giving the application a false API endpoint to query), please can you test the following PR to assist your issue:

```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/2266/head:pr2266
git checkout pr2266
```

To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any options needed>
```

To install the PR, you will need to perform `sudo make install` to install the PR version. 

When running the PR, your version should be: `onedrive v2.4.22-5-gd723491` or greater. An example output is below:
```
(dmd-2.100.2)[alex@onedrive-client-dev onedrive]$ ./onedrive --get-O365-drive-id '*' --verbose
Using 'user' Config Dir: /home/alex/.config/onedrive
Using 'system' Config Dir: /etc/onedrive
No user or system config file found, using application defaults
Using dry-run database copy for OneDrive API query
Checking Application Version ...
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Using Curl defaults for all HTTP operations
Opening the item database ...
All operations will be performed in: /home/alex/OneDrive
Application version: onedrive v2.4.22-5-gd723491
Account Type: business
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 1096605094055
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Initializing the Synchronization Engine ...
Office 365 Library Name Query: *
ERROR: Query of OneDrive for Office 365 Library Name failed
ERROR: Your OneDrive Account and Authentication Scope cannot access this OneDrive API: https://graph.microsoft.com/v1.0/sites?search
ERROR: To resolve, please discuss this issue with whomever supports your OneDrive and SharePoint environment.
```

**Note:** The URL presented will be dynamic, based on your actual MS Graph endpoint being used, which is based on your OneDrive Business Account configuration.

**Reply by rafalcode** [reply]
Thanks very much for looking further into this.

In relation to you first reply, I understand your points.  With `onedrive --get-O365-drive-id '*'`, I kept getting the same error as hugobrehier and, as I got before (when trying out different names), so no change there.

The Onedrive/Sharepoint web interface gives no clues about permissions to the "site library" ... as usual, all the information on the web interface is trivial in nature. In fact, I have a problem distinguishing what is Onedrive and what is Sharepoint, I think there's a big overlap between the two.

That's not a problem I want to burden you with, but just out of interest, my opinion is that Sharepoint seems to be a superset of Onedrive, (analogous to how C++ is a superset to C). It has extra bell and whistles, some of which may be important, but uses Onedrive under the hood.

OK back to the errors, and your second post. Yes, I cloned that branch and got exactly what you are getting, `--display-config` told me I was using version `v2.4.22-6-g6ac4346`.

**Comment by abraunegg** [comment]
@rafalcode 
> OK back to the errors, and your second post. Yes, I cloned that branch and got exactly what you are getting

Thanks for the confirmation on this, and I will merge this PR into 'master'

To resolve this issue - you really need to speak with whomever supports your OneDrive and SharePoint environment. There is nothing that this client can do to resolve this.
